[Merged by Bors] - Add CODING.md

acud opened the pull request:
## Motivation

As a part of an effort to standardize the way we develop software I am suggesting this as an entry point to a discussion about how to write and expect code to be checked into the codebase.
<!-- Give a brief description of the motivation for this PR. 1-2 sentences is fine. -->

## Description

While a lot of the suggestions here don't match what is actually in the codebase, it is meant to describe the point we aspire to get to, and slowly mutate the codebase into that direction by altering what is already there or by the style in which new components are written. It is meant as a departure point into a discussion about the content, rather than an attempt to coerce the content as something set in stone, therefore, suggestions and questions are welcome.
<!-- If applicable please mention the issue addressed by this PR -->
<!-- `Closes #XXXX` links mentioned issues to this PR and automatically closes them when this it's merged -->

<!-- Please describe in detail the changes made. Focus on the reasoning rather than describing the change -->

## Test Plan

Not to achieve a quorum but to see if anyone in the team objects to this format and content.
<!-- Please specify how these changes were tested (e.g. unit tests, manual testing, etc.) -->

acud committed:
chore: add coding.md

fasmat commented on CODING.md:
```diff
@@ -0,0 +1,97 @@
+
+## Concurrency
+
+For every goroutine that is created, one must define how and when it terminates.
```
In most of the code base we use `errgroups` and unless there is a good reason for it any new code shoud in my opinion use `errgroups` too.

acud replied:
So `errgroup` helps with error propagation from goroutines, but it does not define how a goroutine lifecycle should be managed. It has an optional context that can be passed, but finally it is up to the developer to decide how the goroutine lifecycle should be managed. This specific comment is made in order to clarify the conditions that must be met in order for a goroutine to terminate (it might also not even emit an error as of such, as some goroutines are just anonymous functions that return nothing). I agree that we can add that `errgroups` should be used, however, these are two separate issues.

fasmat replied:
This is true but if you take a look at how we usually use `errgroup.Group` you'll see that we are actually not using it for the error propagation or synchronized cancelling in most cases. But rather because

```go
eg.Go(func() error {
   // do stuff
   return nil
})

// Other calls to eg.Go then
eg.Wait()
```
Is syntactically nicer that
```go
var wg sync.Waitgroup
wg.Add(1)
go func () {
   defer wg.Done()
   // do stuff 
}()

// Other go routines with wg.add and wg.done 
wg.Wait()
```

fasmat replied:
Most components also have an `errgroup.Group` that is used to start to routines that do not need to be awaited generally, but will be awaited when shutting down (gracefully exit).

It's a pattern that we started following some time ago and it proved to be useful.

poszu replied:
The `errgroup` is not only nicer to the eye, but also easier to use than waitgroup (can't misplace/forget `wg.Add(1)`).

acud replied:
I totally agree that using `errgroup` is favorable to using `sync.WaitGroup`, however, this is again not what the comment is referring to. The topic is rather a bit more abstract and is mentioned in the [CodeReviewComments](https://go.dev/wiki/CodeReviewComments#goroutine-lifetimes) resource:

> 
> Goroutine Lifetimes
> When you spawn goroutines, make it clear when - or whether - they exit.
> 
> Goroutines can leak by blocking on channel sends or receives: the garbage collector will not terminate a goroutine even if the channels it is blocked on are unreachable.
> 
> Even when goroutines do not leak, leaving them in-flight when they are no longer needed can cause other subtle and hard-to-diagnose problems. Sends on closed channels panic. Modifying still-in-use inputs “after the result isn’t needed” can still lead to data races. And leaving goroutines in-flight for arbitrarily long can lead to unpredictable memory usage.
> 
> Try to keep concurrent code simple enough that goroutine lifetimes are obvious. If that just isn’t feasible, document when and why the goroutines exit.
>

acud replied:
I would like to address all of the specifics on favorable patterns and do's and don't's in another document which would be effectively the style-guide, and keep this one more abstract

fasmat commented on CODING.md:
```diff
@@ -0,0 +1,97 @@
+
+## Error propagation, annotation and handling
+
+Errors must be propagated by the caller function. Error should not be logged and passed at the same time. It is up to the next caller function to decide what should be done in case of an error (e.g. execute specific logic, log, or continue bubbling the error up the stack).
+
```
Should we add to use the general pattern of using switch and `errors.Is`/`errors.As` like:

```go
var customError MyErroType
err := someFunctionCall()
switch {
case errors.Is(err, Error1):
   // handle Error1
case errors.As(err, &customError):
   // handle customError
case err != nil: // another error
   // handle generic error
}
```

or leave that up to the developer?

acud replied:
I think that using `errors.Is/As` should be encouraged yes. About using switches - I'm not 100% sure because it's a general deviation from the usual `if err := ... {}` structure and switches have the problematic aspect of when they have the `default:` case it would handle the happy path which is a bit misleading. Ideally, the happy path should never be nested, but rather handled at the base nesting level of the method (only error paths should be nested). It is indeed sometimes unavoidable to handle happy path this way but switches may cause this to happen more often than not.

Also as a sidenode it should probably also be noted that happy path shouldn't be nested :)

I'd like to cover all those in a separate styleguide that would have examples too.

poszu replied:
Using switch doesn't mean that the happy path must be nested. The errors can be handled in the switch cases and the happy path - below: 
```go
err := someFunctionCall()
switch {
case errors.Is(err, Error1):
   // handle Error1
case err != nil:
   // handle unknown error
}

// happy path goes here
```

fasmat commented on CODING.md:
```diff
@@ -0,0 +1,97 @@
+
+That is, the same problem event can have two log lines, but with different severity. This is the case for a combination of `Error/Debug` or `Warning/Debug`, where `Error` or `Warning` is for the node operator and `Debug` is for the developer to help them investigate the problem.
+
+> The `Error` level should be used with caution when it comes to node operators. It is recommended to log application code errors at the `Error` level if they require node operator intervention or if the node cannot continue its normal operation (unrecoverable errors). Otherwise, recoverable (application code) errors should be logged at the `Debug` level. The `Debug` level should not be used to m ...
+
```
If the `Error` cannot be recovered from then I assume we also terminate the node in a timely manner after printing the `Error` log?

acud replied:
Ideally yes (with the caveat that this comment is made with regards to logging only, not as to how the application itself should behave). I'm not sure if currently we panic on these sort of errors or if there's an actual shutdown mechanism (which is troublesome on its own when systemd or k8s are involved)

fasmat reviewed (commented):
Thanks for providing a great starting point for this! 🙂 

Overall I agree with the suggested coding style. I added a few comments regarding details that I'm unsure about or feel like could be added to the document.

ivan4th commented on CODING.md:
```diff
@@ -0,0 +1,97 @@
+
+Explicit `error` values that are constructed with `errors.New` may have a message prefixed with the package name and a colon character, if it would improve the message clarity.
+
+## Logging Guidelines
```
Re logging: we're currently using `zap` for logging, but also our own `log` facility (`github.com/spacemeshos/go-spacemesh/log`)
The latter is considered deprecated and being phased out, so that the new code should always use `zap` for logging. Maybe that's also worth mentioning

acud committed:
chore: address PR comments

acud committed:
Update CODING.md

acud commented:
I've addressed most of the comments here, please see the diffs that I've pushed. I suggest to address specific style-guide points in a separate document that will be more detailed.

fasmat reviewed (approved):
lgtm, if @ivan4th and @poszu have no objections, I'd say its ready to be merged and go into effect 🙂

acud commented:
bors merge
